opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java

```diff
@@ -97,56 +98,85 @@ public ExprType getExprType() {
+    if (indexMapping == null) {
+      return result;
+    }
+
+    indexMapping.forEach((k, v) -> {
+      var innerMap = (Map<String, Object>)v;
+      // by default, the type is treated as an Object if "type" is not provided
+      var type = ((String) innerMap
+          .getOrDefault(
+              "type",
+              "object"))
+          .replace("_", "");
+      if (!EnumUtils.isValidEnumIgnoreCase(OpenSearchDataType.MappingType.class, type)) {
+        // unknown type, e.g. `alias`
+        // TODO resolve alias reference
+        return;
+      }
+      // create OpenSearchDataType
+      result.put(k, OpenSearchDataType.of(
+          EnumUtils.getEnumIgnoreCase(OpenSearchDataType.MappingType.class, type),
+          innerMap)
+      );
+    });
+    return result;
+  }
+
   /**
    * A constructor function which builds proper `OpenSearchDataType` for given mapping `Type`.
    * @param mappingType A mapping type.
    * @return An instance or inheritor of `OpenSearchDataType`.
    */
-  public static OpenSearchDataType of(MappingType mappingType) {
-    var res = instances.getOrDefault(mappingType.toString(), null);
-    if (res != null) {
-      return res;
-    }
-    ExprCoreType exprCoreType = mappingType.getExprCoreType();
-    if (exprCoreType == ExprCoreType.UNKNOWN) {
-      switch (mappingType) {
+  public static OpenSearchDataType of(MappingType mappingType, Map<String, Object> innerMap) {
+    OpenSearchDataType res = instances.getOrDefault(mappingType.toString(),
+        new OpenSearchDataType(mappingType)
+    );
+    switch (mappingType) {
+      case Object:
+        // TODO: use Object type once it has been added
+      case Nested:
+        if (innerMap.isEmpty()) {
+          return res;
+        }
+        Map<String, OpenSearchDataType> properties =
+            parseMapping((Map<String, Object>) innerMap.getOrDefault("properties", Map.of()));
+        OpenSearchDataType objectDataType = res.cloneEmpty();
+        objectDataType.properties = properties;
+        return objectDataType;
+      case Text:
         // TODO update these 2 below #1038 https://github.com/opensearch-project/sql/issues/1038
-        case Text: return OpenSearchTextType.of();
-        case GeoPoint: return OpenSearchGeoPointType.of();
-        case Binary: return OpenSearchBinaryType.of();
-        case Ip: return OpenSearchIpType.of();
-        default:
-          throw new IllegalArgumentException(mappingType.toString());
-      }
+        Map<String, OpenSearchDataType> fields =
+            parseMapping((Map<String, Object>) innerMap.getOrDefault("fields", Map.of()));
+        return (!fields.isEmpty()) ? OpenSearchTextType.of(fields) : OpenSearchTextType.of();
+      case GeoPoint: return OpenSearchGeoPointType.of();
+      case Binary: return OpenSearchBinaryType.of();
+      case Ip: return OpenSearchIpType.of();
+      case Date:
+        return OpenSearchDateType.create(
+          (String) innerMap.getOrDefault("format", ""));
```

Comment:
I would consider adding a comment that the `""` will use the default date format